core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/node/IModelNodeItem.java

```diff
@@ -30,4 +25,46 @@ public interface IModelNodeItem<D extends IModelDefinition, I extends INamedMode
 
   @Override
   IAssemblyNodeItem getParentContentNodeItem();
+
+  @Override
+  @NonNull
+  default Stream<? extends IModelNodeItem<?, ?>> descendantOrSelf() {
+    return ObjectUtils.notNull(Stream.concat(Stream.of(this), descendant()));
+  }
+
+  @SuppressWarnings("PMD.CompareObjectsWithEquals")
+  @Override
+  @NonNull
+  default Stream<? extends IModelNodeItem<?, ?>> followingSibling() {
+    IModelNodeItem<?, ?> parent = getParentContentNodeItem();
+    return ObjectUtils.notNull(parent == null
+        ? Stream.empty()
+        : parent.modelItems()
+            // need to use != vs !Object.equals to ensure we are matching the same object
+            .dropWhile(item -> this != item)
+            .skip(1));
+  }
+
+  @SuppressWarnings("PMD.CompareObjectsWithEquals")
+  @Override
+  @NonNull
+  default Stream<? extends IModelNodeItem<?, ?>> precedingSibling() {
+    IModelNodeItem<?, ?> parent = getParentContentNodeItem();
+    return ObjectUtils.notNull(parent == null
+        ? Stream.empty()
+        // need to use != vs !Object.equals to ensure we are matching the same object
+        : parent.modelItems().takeWhile(item -> this != item));
+  }
+
+  @Override
+  default Stream<? extends IModelNodeItem<?, ?>> following() {
+    return followingSibling()
+        .flatMap(IModelNodeItem::descendantOrSelf);
+  }
+
+  @Override
+  default Stream<? extends IModelNodeItem<?, ?>> preceding() {
+    return precedingSibling()
+        .flatMap(IModelNodeItem::descendantOrSelf);
+  }
```

Comment:
Non-blocking: I am not understanding how this code works if the calling functions just do a null check but also stream a `Stream.empty()`, but it seems the tests

Reply:
the `precedingSibling` method implements the `preceding-sibling` access. The `descendantOrSelf` method is used to descend into the graph to pick up the rest of the items that are part of the `preceding` axis. `flatMap` replace an item in the stream, with items in the resulting stream generated by the lambda call.